jvm-packages/xgboost4j-spark/src/main/scala/org/apache/spark/SparkParallelismTracker.scala

```diff
@@ -146,22 +148,29 @@ class TaskFailedListener(killSparkContext: Boolean = true) extends SparkListener
 
 object TaskFailedListener {
 
-  var killerStarted = false
+  var killerStarted = new AtomicBoolean(false)
+
+  // mostly for testing
+  val sparkContextShutdownLock = new AnyRef
 
   private def startedSparkContextKiller(): Unit = this.synchronized {
-    if (!killerStarted) {
+    if (!killerStarted.get()) {
       // Spark does not allow ListenerThread to shutdown SparkContext so that we have to do it
       // in a separate thread
       val sparkContextKiller = new Thread() {
         override def run(): Unit = {
           LiveListenerBus.withinListenerThread.withValue(false) {
             SparkContext.getOrCreate().stop()
+            sparkContextShutdownLock.synchronized {
+              killerStarted.set(false)
```

Comment:
I don't know whether there is a situation triggerring multiple sparkContextKiller?

one task failure will trigger 1 sparkContextKiller, and after killerStarted.set(false). another task failure will trigger another sparkContextKiller ?

Reply:
it will not, when we reach here, it means that we have returned from SparkContext.stop() which cleaned all msgs (including TaskFailedEvent)...in the other word, when we want to set this to false, all attempts of starting sparkContextKiller should have been gone through...in the successive attempts, since we haven't set this to false, they will not start new killers